### ISSUE TITLE
Registered C++ member function not getting called from Lua usertype

### ISSUE THREAD

**Issue by gitbetter** [ThePhD/sol2#796]
Hello,

I am trying to call a C++ member function from Lua code but I get the common "received 'nil' for self argument" error, but I can't seem to find anything wrong with the code. The member function gets registered properly for the usertype and, after printing some table values from the Lua script, it seems that the function is present, so I'm not sure what it could be.

My usertype looks as follows

```c++
lua.new_usertype<ScriptableProcess>("ScriptableProcess", 
    "Create", sol::factories(ScriptableProcess::CreateFromScript),
    "Initialize", &Process::Initialize,
    "Update", &Process::Update,
    "Finish", &Process::Finish,
    "Fail", &Process::Fail,
    "Pause", &Process::Pause,
    "Continue", &Process::Continue,
    "Abort", &Process::Abort,
    "IsAlive", &ScriptableProcess::IsAlive,
    "IsDead", &ScriptableProcess::IsDead,
    "IsPaused", &ScriptableProcess::IsPaused,
    "AttachChild", &ScriptableProcess::ScriptAttachChild,
    sol::base_classes, sol::bases<Process>()
  );
```

The function in question, which produces the error described, is the Finish member function. Here's the Lua script I am running.

```lua
TestScriptProcess = ScriptableProcess;

function TestScriptProcess:OnInitialize()
	print("Initializing...");
	self.count = 0;
end

function TestScriptProcess:OnUpdate()
	self.count = self.count + 1;
	print("Count: " .. self.count);
	if (self.count >= 10) then
		self:Finish();    -- Error happens here, "received 'nil' for self argument..."
	end
end

function TestScriptProcess:OnFinish()
	print("Done!");
end

parent = TestScriptProcess:Create({frequency = 1000.0});
child = TestScriptProcess:Create({frequency = 500.0});
parent:AttachChild(child);
AttachProcess(parent);
```
... and here is the simple Finish function as implemented in the Process base class

```c++
virtual void Finish() { state_ = ProcessState::Finished; OnFinish(); };
```

The registered functions that are part of the derived class seem to be working, and the OnInitialize, OnUpdate, etc. functions are callbacks that get set and called from the C++ environment. I am running macOS Mojave with no comprehensive IDE. Currently using Visual Studio Code and compiling and linking the necessary libraries from a bare bones Makefile.

I looked at all the sol2 examples I could and I can't find anything wrong with my code...Registered the base class and all.

Any help would be highly regarded, and if I left any crucial piece of info or code out let me know.

**Comment by gitbetter** [comment]
Ok I managed to figure out what the issue was. It stemmed from my factory Create function (ScriptableProcess::CreateFromScript), which previously looked like this

```c++
std::shared_ptr<ZScriptableProcess> ZScriptableProcess::CreateFromScript(sol::table self, sol::table constructionData) {
  std::shared_ptr<ZScriptableProcess> process(new ZScriptableProcess);
  process->self_ = self;
  if (!process->BuildCppDataFromScript(self, constructionData)) {
    process->self_ = sol::lua_nil;
  }
  return process;
}
```
I was setting the self_ variable of the process to a sol::table type, and then passing that self_ variable to the saved Lua functions, as in `scriptUpdate_(self_);`. Turns out that since I am working with a usertype, which handles the metatable on its own, the table I was saving to self_ did not include the proper metatable metafunctions and so no _self_ existed. To fix it I just passed the actual instance to the function, as in `scriptUpdate_(*this);` and it worked as expected.
